### PR TITLE
Add confusion matrix

### PR DESCRIPTION
**Pull request by feifjiang**
This PR replaces #532 #530

### CODE REVIEW

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala [hunk 1]
```diff
@@ -133,20 +168,176 @@ private[op] class OpMultiClassificationEvaluator
...
+    val confusionMatrixMap = labelPredictionCtRDD.map {
+      case ((label, prediction, _), count) => ((label, prediction), count)
+    }.reduceByKey(_ + _)
+      .map {
+        case ((label, prediction), count) => (label, prediction) -> count
+      }.collectAsMap()
+
+    cmClasses.map(label => {
+      cmClasses.map(prediction => {
+        confusionMatrixMap.get((label, prediction)).getOrElse(0L)
```
Comment (gerashegalov) [inline core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:212]: there is syntactic sugar for this 
```scala
confusionMatrixMap.getOrElse((label, prediction), 0L)
```

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala [hunk 2]
```diff
@@ -133,20 +168,176 @@ private[op] class OpMultiClassificationEvaluator
...
+      })
+    }).flatten
```
Comment (gerashegalov) [inline core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:214]: flatten of a map means we want a `flatMap` on Line 210 or we can use the for comprehension
```
    for {
      label <- cmClasses
      prediction <- cmClasses
    } yield {
      confusionMatrixMap.getOrElse((label, prediction), 0L)
    }
```

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala [hunk 3]
```diff
@@ -133,20 +168,176 @@ private[op] class OpMultiClassificationEvaluator
...
+ * @param data RDD of (label, prediction, prediction probability vector)
+ * @return sequence of Matrix that corresponds to the confMatrixThresholds
+*/
+
+  def calculateConfMatrixMetricsByThreshold(
+    data: RDD[(Double, Double, Array[Double])]): MulticlassConfMatrixMetricsByThreshold = {
+
+    val labelCountsRDD = data.map { case (label, _, _) => (label, 1L) }.reduceByKey(_ + _)
+    val cmClasses = labelCountsRDD.sortBy(-_._2).map(_._1).take($(confMatrixNumClasses)).toSeq
+    val cmClassesSet = new HashSet() ++ cmClasses
```
Comment (gerashegalov) [inline core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:243]: Let us do
```
val cmClassesSet = cmClasses.toSet
```

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala [hunk 4]
```diff
@@ -133,20 +168,176 @@ private[op] class OpMultiClassificationEvaluator
...
+
+    val dataTopNLabels = data.filter { case (label, _, _) => cmClassesSet.contains(label) }
```
Comment (gerashegalov) [inline core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:245]: Is this equivalent to?
```
    val dataTopNLabels = data.filter { case (label, prediction, _) => 
      cmClassesSet.contains(label) || cmClassesSet.contains(prediction)
    }
```

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala [hunk 5]
```diff
@@ -133,20 +168,176 @@ private[op] class OpMultiClassificationEvaluator
...
+      case ((label, prediction), count) => (label, (prediction, count))
+    }.groupByKey()
+      .map { case (label, predictionCountsIter) => {
+        val misClassificationCtMap = predictionCountsIter.toSeq
+          .filter { case (pred, _) => pred != label }
+          .sortBy(-_._2)
+          .take($(confMatrixMinSupport)).toMap
+
+        val labelCount = predictionCountsIter.map(_._2).reduce(_ + _)
+        val correctCount = predictionCountsIter.filter {case (pred, _) => pred == label }
```
Comment (gerashegalov) [inline core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:296]: filter + map usually calls for a single collect
```
        val correctCount = predictionCountsIter
          .collect { case (pred, count) if pred == label => count }
          .reduceOption(_ + _)
          .getOrElse(0L)
```

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala [hunk 6]
```diff
@@ -133,20 +168,176 @@ private[op] class OpMultiClassificationEvaluator
...
+
+        MisClassificationsPerCategory(
+          category = prediction,
+          totalCount = predictionCount,
+          correctCount = correctCount,
+          misClassifications = sortedMisclassificationCt
+        )
+      }
+    }.collect()
+      .sortBy(-_.totalCount).toSeq
```
Comment (gerashegalov) [inline core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:332]: can we reverse sortBy and collect to give Spark more work? 

toSeq is redundant

**Review by gerashegalov — CHANGES_REQUESTED** [review]

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala (hunk 4 above)
Comment (feifjiang) [reply core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:245]: It equals to &&. I like this simpler syntax - will change it

**Review by gerashegalov — APPROVED** [review]
LGTM, question re confMatrixThresholds parameter upper bound

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala [hunk 7]
```diff
@@ -133,20 +166,173 @@ private[op] class OpMultiClassificationEvaluator
...
   }
 
+/**
+ * function to construct the confusion matrix for the top n most occurring labels
+ * @param labelPredictionCtRDD RDD of ((label, prediction, confidence), count)
+ * @param cmClasses the top n most occurring classes, sorted by counts in descending order
+ * @return an array of counts
+ */
+  def constructConfusionMatrix(
+    labelPredictionCtRDD: RDD[((Double, Double, Double), Long)],
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:198]: Why not define a `case class LabelPredConf(label: Double, prediction: confidence: Double, count: Long)`?

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala [hunk 8]
```diff
@@ -133,20 +166,173 @@ private[op] class OpMultiClassificationEvaluator
...
+
+    for {
+      label <- cmClasses
+      prediction <- cmClasses
+    } yield {
+      confusionMatrixMap.getOrElse((label, prediction), 0L)
+    }
+  }
+
+  object SearchHelper extends Serializable{
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:213]: `private[evaluators] SearchHelper extends Serializable {`

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala [hunk 9]
```diff
@@ -133,20 +166,173 @@ private[op] class OpMultiClassificationEvaluator
...
+
+    def findThreshold(arr: IndexedSeq[Double], element: Double): Double = {
+      if (element > arr.last) arr.last
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:216]: 1. please check that arr is non empty
2. use arr.head instead of arr(0)

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala [hunk 10]
```diff
@@ -133,20 +166,173 @@ private[op] class OpMultiClassificationEvaluator
...
+
+    val labelPredictionCountRDD = data.map {
+      case (label, prediction) => ((label, prediction), 1L) }
+      .reduceByKey(_ + _)
+
+    val misClassificationsByLabel = labelPredictionCountRDD.map {
+      case ((label, prediction), count) => (label, (prediction, count))
+    }.groupByKey()
+      .map { case (label, predictionCountsIter) => {
+        val misClassificationCtMap = predictionCountsIter.toSeq
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:287]: Here and below - instead of iterating through predictionCountsIter twice for cases where `pred != label` and `pred == label`, better use `predictionCountsIter.partition(...)`

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala [hunk 11]
```diff
@@ -133,20 +166,173 @@ private[op] class OpMultiClassificationEvaluator
...
+          misClassifications = misClassificationCtMap
+        )
+      }
+    }.sortBy(-_.totalCount).collect()
+
+    val misClassificationsByPrediction = labelPredictionCountRDD.map {
+      case ((label, prediction), count) => (prediction, (label, count))
+    }.groupByKey()
+      .map { case (prediction, labelCountsIter) => {
+        val sortedMisclassificationCt = labelCountsIter.toSeq
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:310]: This code block is almost identical to the one in misClassificationsByLabel. Please introduce a local helper function to DRY it.

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala [hunk 12]
```diff
@@ -352,6 +539,54 @@ case class MultiClassificationMetricsTopK
...
+ */
+case class MisClassificationMetrics
+(
+  ConfMatrixMinSupport: Int,
+  MisClassificationsByLabel: Seq[MisClassificationsPerCategory],
+  MisClassificationsByPrediction: Seq[MisClassificationsPerCategory]
+)
+
+/**
+ * Case class that stores the most frequently mis-classified classes for each label/prediction category
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:573]: replace `Case class that stores` with `Container to store`

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala [hunk 13]
```diff
@@ -133,20 +166,173 @@ private[op] class OpMultiClassificationEvaluator
...
+        else arr(insertionPoint-1)
+      }
+    }
+  }
+
+/**
+ * function to calculate confusion matrix for TopK most occurring labels by confidence threshold
+ *
+ * @param data RDD of (label, prediction, prediction probability vector)
+ * @return a MulticlassConfMatrixMetricsByThreshold object
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:231]: `@return a MulticlassConfMatrixMetricsByThreshold instance`

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala [hunk 14]
```diff
@@ -352,6 +539,54 @@ case class MultiClassificationMetricsTopK
...
+ * Metrics for multi-class confusion matrix. It captures confusion matrix of records of which
+ * 1) the labels belong to the top n most occurring classes (n = confMatrixNumClasses)
+ * 2) the top predicted probability exceeds a certain threshold in confMatrixThresholds
+ *
+ * @param confMatrixNumClasses value of the top n most occurring classes in the dataset
+ * @param confMatrixClassIndices label index of the top n most occuring classes
+ * @param confMatrixThresholds a sequence of thresholds
+ * @param confMatrices a Matrix object that stores the confusion matrix for each threshold in confMatrixThresholds
+ */
+case class MulticlassConfMatrixMetricsByThreshold
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:552]: 1. Prefixing everything with `contMatrix` seems redundant 
2. Why is everything lowercased here? Why not:
```scala
case class MulticlassConfMatrixMetricsByThreshold
(
  NumClasses: Int,
  ClassIndices: Seq[Double],
  Thresholds: Seq[Double],
  ConfMatrices: Seq[Seq[Long]]
) extends EvaluationMetrics
```

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala [hunk 15]
```diff
@@ -352,6 +539,54 @@ case class MultiClassificationMetricsTopK
...
+ *
+ * @param category a category which a record's label or prediction equals to
+ * @param totalCount total # of records in that category
+ * @param correctCount # of correctly predicted records in that category
+ * @param misClassifications the top n most frequently misclassified classes (n = confMatrixMinSupport) and
+ *                           their respective counts in that category. Ordered by counts in descending order.
+ */
+case class MisClassificationsPerCategory
+(
+  category: Double,
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:583]: same here - shouldn't the members be capitalized?

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala [hunk 16]
```diff
@@ -352,6 +539,54 @@ case class MultiClassificationMetricsTopK
...
+ * 2) the top predicted probability exceeds a certain threshold in confMatrixThresholds
+ *
+ * @param confMatrixNumClasses value of the top n most occurring classes in the dataset
+ * @param confMatrixClassIndices label index of the top n most occuring classes
+ * @param confMatrixThresholds a sequence of thresholds
+ * @param confMatrices a Matrix object that stores the confusion matrix for each threshold in confMatrixThresholds
+ */
+case class MulticlassConfMatrixMetricsByThreshold
+(
+  confMatrixNumClasses: Int,
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:554]: Don't you need to add `@JsonDeserialize` annotations here?

File: local/src/test/scala/com/salesforce/op/local/OpWorkflowModelLocalTest.scala [hunk 17]
```diff
@@ -123,7 +123,7 @@ class OpWorkflowModelLocalTest extends FlatSpec with TestSparkContext with TempD
     }
     log.info(s"Scored ${expectedScores.length * numOfRuns} records in ${elapsed}ms")
     log.info(s"Average time per record: ${elapsed.toDouble / (expectedScores.length * numOfRuns)}ms")
-    elapsed should be <= 25000L
+    elapsed should be <= 50000L
```
Comment (tovbinm) [inline local/src/test/scala/com/salesforce/op/local/OpWorkflowModelLocalTest.scala:126]: This change is not related. Let's omit this from this PR.

Comment (feifjiang) [reply local/src/test/scala/com/salesforce/op/local/OpWorkflowModelLocalTest.scala:126]: @tovbinm  this is a flapper test that sometimes fails with 25000L as the threshold. That's why I increased it to 50000L

Comment (tovbinm) [reply local/src/test/scala/com/salesforce/op/local/OpWorkflowModelLocalTest.scala:126]: Let's not double it. Let's increase it a little, e.g. 35000

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala [hunk 18]
```diff
@@ -122,9 +120,9 @@ private[op] class OpMultiClassificationEvaluator
     parent = this,
     name = "confMatrixThresholds",
     doc = "sequence of threshold values used for confusion matrix metrics",
-    isValid = _.forall(x => x >= 0.0 && x <= 1.0)
+    isValid = _.forall(x => x >= 0.0 && x < 1.0)
```
[the remark this answers is not in the document]
Comment (feifjiang) [reply core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:123]: For multinomial logistic regression, the scenario where top probability equals to 1.0 will never happen. For random forest it's quite unlikely, unless the record falls onto one and the only one leaf node of the same class in every tree. Therefore it doesn't really make sense to have confidence threshold of 1.0

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala (hunk 7 above)
Comment (feifjiang) [reply core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:198]: @tovbinm this case class is only used within this class for confusion matrix calculation and not sent outbound, so I didn't explicitly define it. In a future PR we can use Dataframe APIs/Row object instead of RDD tuples (suggestion from @gerashegalov ) in OpMultiClassificationEvaluator.scala, so that it's more explicit than the tuples and we don't have to declare it every time in the anonymous functions

Comment (tovbinm) [reply core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:198]: this can be a private case class for this module.

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala (hunk 14 above)
Comment (feifjiang) [reply core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:552]: 1. There's already a `thresholds` defined for the threshold metrics, so adding a `confMatrix` prefix is necessary to avoid confusion. Also `confMatrixNumClasses` and `confMatrixClassIndices` specifically refer to the cutoff classes for confusion matrix calculation, so I think it's better to be explicit about them. Otherwise `numClasses` could mean number of classes for other metrics or for all the data.

2. Yes I can change all the element to uppercase. There doesn't seem to be a consistency of whether to use uppercase or camel case in the case classes that extends `EvaluationMetrics` here.

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala (hunk 10 above)
Comment (feifjiang) [reply core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:287]: @tovbinm  followed gera's suggestion and changed it to  `.collect { case (pred, count) if pred == label => count }`

Comment (tovbinm) [reply core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:287]: ok.

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala (hunk 16 above)
Comment (feifjiang) [reply core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:554]: This case class will be further transformed in E1 before saving the metrics, so there doesn't seem to be a need to add an annotation here. The JsonDeserialize tag was added to the Map as `ModelSelectorSummaryTest` will break without it.

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala (hunk 18 above)
[the remark this answers is not in the document]
Comment (gerashegalov) [reply core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:123]: just wondering how much value is there to restrict it here?

[the remark this answers is not in the document]
Comment (feifjiang) [reply core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:123]: @gerashegalov are you concerned about too many values present in the `confMatrixThresholds` array? That's a valid concern, I can add a check for that. It's a

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala (hunk 16 above)
Comment (tovbinm) [reply core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala]: I think it's still a good idea to add deser annotations, in case someone would need to parse json into classes later on.

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala (hunk 14 above)
Comment (feifjiang) [reply core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:563]: @tovbinm  Changed case class data fields name to upper case. Please see updated PR

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala (hunk 15 above)
Comment (feifjiang) [reply core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala]: @tovbinm changed the members to capitalized.

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala (hunk 16 above)
Comment (feifjiang) [reply core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala]: @tovbinm added @JsonDeserialize annotation to both `ConfMatrixClassInidces` and `ConfMatrixThresholds` in the updated PR.

**Review by tovbinm — APPROVED** [review]

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala [hunk 19]
```diff
@@ -597,6 +600,13 @@ case class MisClassificationsPerCategory
   MisClassifications: Map[Double, Long]
 )
 
+case class labelPredictionConfidence
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:603]: @feifjiang are you planning to use LabelPredictionConfidenceCt and labelPredictionConfidence somewhere?

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala (hunk 7 above)
Comment (feifjiang) [reply core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:200]: If defining a case class `LabelPredConf(label: Double, prediction: confidence: Double, count: Long)`, it would require an additional map function to map the key-val pairs to this case class after the ```    val labelPredictionConfidenceCountRDD = dataTopNLabels.map{
      case (label, prediction, proba) => {
        ( (label, prediction, SearchHelper.findThreshold(sortedThresholds, proba.max)), 1L )
      }
    }.reduceByKey(_ + _) ```

And in then in `constructConfusionMatrix`, it will be `    val confusionMatrixMap = labelPredictionCtRDD.map( e => ((e.Label, e.Prediction), e.Count)).reduceByKey(_ + _).collectAsMap() `

Comment (tovbinm) [reply core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:200]: It's not mandatory. If it only scoped within one function / file - no need to create a case class.

Comment (feifjiang) [reply core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:200]: If defining a case class here, the new constructConfusionMatrix would look like 

 ```
 def constructConfusionMatrix(
    labelPredictionCtRDD: RDD[LabelPredictionConfidenceCt],
    cmClasses: Seq[Double]): Seq[Long] = {

    val confusionMatrixMap = labelPredictionCtRDD.map( r => ((r.Label, r.Prediction), r.Count))
      .reduceByKey(_ + _).collectAsMap()
```
And an additional map function will be needed  after `reduceByKey` in 

```
    val labelPredictionConfidenceCountRDD = dataTopNLabels.map{
      case (label, prediction, proba) => {
        ( (label, prediction, SearchHelper.findThreshold(sortedThresholds, proba.max)), 1L )
      }
    }.reduceByKey(_ + _)
      .map {
        case ((label, pred, confidence), ct) => LabelPredictionConfidenceCt(label, pred, confidence, ct)
      }
```
This case class is only used twice in the entire class. It doesn't seem to make the code more succinct. What would be the benefit of doing it? @tovbinm

File: core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala (hunk 19 above)
Comment (feifjiang) [reply core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala:601]: @tovbinm  sorry - forgot to remove it. Just updated the PR again.

**Comment by feifjiang** [comment]
@tovbinm @gerashegalov Can either one of you merge my PR with master? I don't have write access.

**Comment by tovbinm** [comment]
@feifjiang thank you for your contribution!

**Comment by feifjiang** [comment]
@tovbinm  Thank you for the comments and being so responsive!!
